[Woo POS] [Cash & Receipts] Handle navigation when order completed via cash

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #14602

## Description
This PR addresses POS navigation when we attempt to collect a cash payment for an order:
- Upon success, the success view is presented, and we can proceed to send a receipt or start a new order.
- Upon failure, an error message is presented, and we remain in the cash collection view.

Currently we're using the `PointOfSaleCardPresentPaymentSuccessMessageView` for both card payments and cash payments success scenarios, and the different `PointOfSaleCardPresentPaymentMessageType` cases are tied to the card payment state. 

I've approached this without touching the card payment state or the card service but by introducing a property to react alongside the card payment messages and leaving the completion handling to the app side, as the POS does nothing regarding cash. Happy to iterate as needed, as another option could be to use some sort of `PaymentResult` state to track both card or cash state as needed.

## Testing
- Enable the `.acceptCashForPointOfSale` flag
- Add items to the cart > checkout > tap "Cash payment"
- Add any value to the textfield, or none, tap on "Mark order as completed"
- Observe that the success view is presented, and we can either send a receipt or start a new order
- Repeat the process by throwing an error on:
```
    private func markComplete() async throws {
        try await posModel.collectCashPayment()
    }
```
- Observe that the error message is presented, and disappears once we start typing any value in the text field

### Error:

https://github.com/user-attachments/assets/02c8f446-e8a4-43a4-8901-35c8194d7a04

### Success:

https://github.com/user-attachments/assets/4edb7cf4-d921-4b6e-9306-38f4ed85aaa3


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
present payment success on cash completion

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Simplify callback

iamgabrielma committed:
lint

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/TotalsView.swift:
```diff
@@ -88,13 +89,23 @@ struct TotalsView: View {
                                            in: totalsFieldAnimation)
             }
         }
+        .fullScreenCover(isPresented: $shouldShowCollectCashPayment) {
```
I'm not very happy with this duplication here, trying to merge both views under a common type doesn't seem to respond well to changes, specially when dismissing we don't seem to be able to go from the success view to starting a new order, or keeping them instantiated below other views. For reference I used something like the following:
```
// A common type for which view we'll be presenting as .fullScreenCover:
enum FullScreenCoverType: Identifiable {
    case collectCash(total: PointOfSaleOrderTotals)
    case paymentSuccess(total: PointOfSaleOrderTotals)

// With a common state, or nil for when there's nothing to present:
@State private var activeFullScreenCover: FullScreenCoverType?

// and we present one of the other when pressing the collection button or through internal state:
if case .loaded(let total) = posModel.orderState {
        activeFullScreenCover = .collectCash(total: total.orderTotal)
    }

// or 
activeFullScreenCover = .paymentSuccess(total: total.orderTotal)
```

joshheald replied:
Yeah, this is a bit strange. Having them in separate presentations is bound to make our life difficult when we want to make the transitions correct though.

My question is whether the `fullscreenCover` is the correct approach here.

We're asking the merchant to _choose between_ cash and card payment. Both choices can lead to a successful payment. The card option is emphasised, but once the user makes a choice, either has equal weight/importance.

The choice to make it a card payment is made when the shopper taps their card. At that point, we push the new, full-screen "Processing" view into place and focus on the card payment.

For Cash, it might be appropriate to do an equivalent push when the merchant chooses that payment method, by tapping the `Cash payment` button.

## Using `fullscreenCover`

If we're using a fullscreen cover for this, we should show the success screen in that same view. The `New order` button should first reset the underlying POS state, then dismiss the cover. That would be consistent use of the presentation approach – Cash takes you out of the main card payment flow, then you tap "Mark complete", and get the success or error all within your branch. If you close it before marking complete, you go back to the same place in card payment. If you tap `New order`, you don't see the POS rearranging itself underneath, it happens before the cash flow is dismissed.

You may need to think carefully about what "reset the underlying POS state" means – e.g. once you set the `orderStage = .building` in the `PointOfSaleAggregateModel`, as happens currently, the `TotalsView` will no longer be drawn, which causes the cash payment `fullscreenCover` to be dismissed immediately. The transition from the TotalsView back to editing works well by changing that state... but not so much when the success view is shown in a fullscreen cover.

With the fullscreenCover approach, we don't neccesarily need to use the same view for success. We could, but it might be simpler to have a Cash-specific success view. However, from my initial look at the code I think you'd still have to move the presentation of the fullscreenCover to the dashboard, otherwise it will auto-dismiss when you set the state back to `building`. Perhaps it's possible to tweak this with matched geometries, but I suspect not.

I was able to get it working to neatly present the success view this way... but it doesn't dismiss all that nicely because I didn't make any changes to the reset code.

https://github.com/user-attachments/assets/1bc138da-2af3-4f4e-abdf-3df54f7219c8

Code for this (in `TotalsView`):

```
        .fullScreenCover(isPresented: $shouldShowCollectCashPayment) {
            if case .loaded(let total) = posModel.orderState {
                NavigationStack {
                    PointOfSaleCollectCashView(orderTotal: total.orderTotal, onCashPaymentSuccess: {
                        shouldShowPaymentSuccessView = true
                    })
                    .matchedGeometryEffect(id: Constants.matchedGeometryCashId,
                                           in: totalsFieldAnimation)
                    .navigationDestination(isPresented: $shouldShowPaymentSuccessView) {
                        let viewModel = PointOfSaleCardPresentPaymentSuccessMessageViewModel(formattedOrderTotal: total.orderTotal)
                        PointOfSaleCardPresentPaymentInLineMessage(messageType: .paymentSuccess(viewModel: viewModel))
                            .matchedGeometryEffect(id: Constants.matchedGeometryCashId,
                                                   in: totalsFieldAnimation)
                        .navigationBarHidden(true)
                    }
                    .navigationBarHidden(true)
                }
            }
        }
```

Note that I wouldn't put all this in TotalsView for real code – it should be possible to do it within the CollectCashView, so you don't need the `onCashPaymentSuccess` closure.

## Using an equivalent push to Card payments

I think we can improve the UX of the POS by keeping the navigation approaches for payments equivalent. We also won't need to do any special handling of the POS reset if we do it this way.

When the shopper taps a card, we would push the fullscreen `Processing` view. When the merchant taps `Cash payment`, we would push the fullscreen `Cash payment` view, allowing them to enter the amount and tap `Mark complete`.

Either way, when the payment action is successful, we push the `Order complete` view, and allow receipts to be printed and the POS to be reset with the `New order` button.

That's the theory. Unfortunately, it would need a little rework in the `TotalsView` code, which currently expects card payments only.

At the moment, it assumes that after the order is loaded, it's usually showing the `cardReaderView`, which incudes all card payment states (fullscreen and part) and the connect your reader prompt. It's often showing the totals fields, and it depends on the (card)`paymentState` to decide whether or not the overall `TotalsView` is fullscreen or not, which is how the "push" to payment processing is shown.

I've not done a deep dive on this, and it would need exploration, but I _think_ we could keep the `cardReaderView` view builder, but potentially have it show the cash payment screen, as a fullscreen view.

To do it well, we would want to change `PointOfSalePaymentState`. I think it would be best as a nested state, demoting the existing enum to be `PointOfSaleCardPaymentState`, adding an equivalent for cash, and then having a case with an associated value for each at the top level. That can get unpleasant to deal with, so a struct might also be appropriate. Or you could try adding `acceptingCash` and `cashPaymentSuccessful` to the existing enum, along with a new convenience init. 

My concern about adding cash-related states is the whole thing where the card payment service expects to update that state as events happen, and those events might happen right after the merchant taps the `Cash payment` button. On one hand – accepting a cash payment ought to be a branch in our flow, and we should cancel the card payment first and make that impossible... on the other, the Terminal SDK could end up doing stuff when we don't expect it, and if we ignore those updates and get out of sync with the service, that gets bad too.

This second option isn't well explored yet, and I'm happy to do that with you if you like. Let me know if/how you want to proceed... it'll need some more thought and planning right now, but I think lead to a better experience and more maintainable code in the future.

joshheald commented:
From initial testing, it works well, but various questions which are mostly about me orienting myself in the project after AFK:

1. The amount typed in has no bearing on whether you can accept the cash payment or not – I'm assuming this is a known issue?
2. The payment amount is always shown as the order total, regardless of what cash amount is typed in.
3. There's no transition between the cash payment and order complete screen – is this tracked in another issue?
4. Cash breakdown isn't added to the order notse as it can be from the order creation flow.

Fine to merge with any of the above as it's feature flagged.

I'll carry on with the review now.

joshheald commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -153,6 +150,11 @@ private extension PointOfSaleCollectCashView {
         )
+        static let failedToCollectCashPayment = NSLocalizedString(
+            "pointOfSale.cashview.failedToCollectCashPayment",
+            value: "Error trying to process payment. Try again.",
```
Do we expect this to be the final error text for all errors? (I'd suggest not... but in that case it might be better to avoid merging with the key as it is now, or we just end up burning that key. You could append `.draft`, perhaps?) 

What's the potential causes of the error; does a retry make sense?

Presumably we'll have validation errors as well, will they be shown in the same place?

iamgabrielma replied:
At the moment I think this is the only case where we should present this string, so it should be fine, as we only land in this catch block when we're unable to update the order. For amount validation errors we should be updating the string directly when we perform `validateAmount()` (not implemented yet, to be done here: https://github.com/woocommerce/woocommerce-ios/issues/14749 )

That said it's better to be safe than sorry, I've added a `.draft` on https://github.com/woocommerce/woocommerce-ios/pull/14795/commits/9bbd5a3c9ac904acfea91323eaaa90896328bc48 until we know for sure.

joshheald reviewed (commented):
I've made a couple of suggestions, but I agree that having two `fullscreenCover` instantiations for a single flow isn't what we want...

iamgabrielma committed:
Update payment state enum with associated values

The top level enum for POS Payment State now has 2 options: card or cash, and the specific payment state is known through their associated values

iamgabrielma committed:
trigger cash collection state and cash view

We no longer use fullscreencover to present the cash and success view but rely on the cash payment state, sharing the flow with card payment state

iamgabrielma committed:
remove callback, success view on state change

iamgabrielma committed:
make test target compile

iamgabrielma committed:
Tests for cash collection payment state change

iamgabrielma committed:
update localization key while work in progress

iamgabrielma committed:
Dismissing cash returns to cart building

iamgabrielma commented:
Thanks for the review and your input on how this should be handled @joshheald , greatly appreciated 🙇 

> The amount typed in has no bearing on whether you can accept the cash payment or not – I'm assuming this is a known issue?
> The payment amount is always shown as the order total, regardless of what cash amount is typed in.

Yes, no validation around the amount or order total is done yet, any value (or no value) passes the check. Will be tackled here: https://github.com/woocommerce/woocommerce-ios/issues/14749

> There's no transition between the cash payment and order complete screen – is this tracked in another issue?

Is not logged yet, added here: https://github.com/woocommerce/woocommerce-ios/issues/14808

> Cash breakdown isn't added to the order notse as it can be from the order creation flow.

Oh! I didn't know this was a thing, I see in the app that we have a specific "Record transaction details in order note" toggle to log those or not, I logged this separately here https://github.com/woocommerce/woocommerce-ios/issues/14809 and will check with Android as well.

### fullscreenCover vs push payment state

Thanks for the guidance here, I've tested a bit the fullScreenCover presentation but as you mention it feels less solid and more prone to future issues than pushing payment state and treating both cash and card payments with the same priority. In https://github.com/woocommerce/woocommerce-ios/pull/14795/commits/25771f2bae440067b0c14f8b7e83587c06e28131 we make payment state to have only 2 cases, cash or card, and the state is handled via the associated values. 

As you mention is a bit more cumbersome to handle the cases, but I don't feel it's a big deal either, as also makes them clearer and more granular. On https://github.com/woocommerce/woocommerce-ios/pull/14795/commits/de82cd1fd5617968c867c2f805cf0538b23d1246 we let the system know that we're collecting cash or that cash has been successfully collected and then what should be presented is handled automatically.

I'm also interested in any thoughts you might have about returning to the cart building stage upon we cancel/dismiss the cash view. Here: https://github.com/woocommerce/woocommerce-ios/pull/14795/commits/32a7d2994a11880b697e8b1577cfcbd6787c0e8e I've went ahead with this for simplicity but I haven't dig deeper into what should be the card payment state if we cancel cash collection, so it felt safer to do it like this for now.

There's a few details to handle still, I can handle those in this PR or in a follow-up for simplicity. For example:

- Some empty spacing appears in the bottom when collecting cash, most likely the spacers and padding when we were assuming only card payments. I can tackle this when also removing the floating card reader button.
- Rename card payment layout to payment layout or similar
- Additional tests for cash states
- Extract the paymentView into smaller components

https://github.com/user-attachments/assets/531c30fe-c6e8-47c8-9918-b0d2648cd4e8

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -199,6 +199,14 @@ extension PointOfSaleAggregateModel {
+
     @MainActor
     func collectCashPayment() async throws {
         try await orderController.collectCashPayment()
```
`cashPaymentSuccess()` should probably be triggered here, instead of from the view.

iamgabrielma replied:
You're totally right, updated on https://github.com/woocommerce/woocommerce-ios/pull/14795/commits/9e1a5c43094f1639ddbeef0ab6ff4dceb126610f

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -199,6 +199,14 @@ extension PointOfSaleAggregateModel {
+        paymentState = .cash(.collectingCash)
+    }
+
+    func cashPaymentSuccess() {
```
It feels a bit strange for this to be a non-private function.

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -199,6 +199,14 @@ extension PointOfSaleAggregateModel {
     }
 
+    func startCashPayment() {
+        paymentState = .cash(.collectingCash)
```
Setting the payment state is always risky – the card payment service can just change it. We'll have to do it for this approach to work... but while `PointOfSaleAggregateModel` is the source of truth for the payment state, it can be mutated by outside events.

We should cancel the prepared card payment here, and ideally ignore/silence card reader events that happen until the cash payment is completed, or cancelled. Note that cancelling a payment is not instant, and might fail (this is unlikely, unless the shopper has tapped their card.)

Currently, the card reader is still listening for payments in the background, and it's possible to pay for an order twice – cash and card – if you tap the card right after the cash payment is made. See the video; I checked and the card payment did go through, even though the order is paid by cash.


https://github.com/user-attachments/assets/8cf18e64-6da3-4f70-8b02-6812ff8735a3

iamgabrielma replied:
I might need some further guidance here to make this more solid. I've been trying what would be the best way to handle disallowing card payments while cash collection is ongoing, and the closer I got is here: 2ee61638f2a043b34754f424450532524703d770

- Starting to collect cash triggers `isCashPaymentInProgress`
- When new payment events go through, if the payment event is of type `.card` and `isCashPaymentInProgress` is true, then we cancel it.
- If cash collection is cancelled, we re-enable card payment events via `startPaymentWhenCardReaderConnected()` so the user can just tap without further setup
- If cash collection is successful, we just trigger back `isCashPaymentInProgress` to false.

While it seems to work 90% of the time, I still see some strange behaviour with the payment state sometimes when trying to collect cash and tapping the card in the reader:
- At times, the view has gone blank when the cash view is rendered and we tap the card reader, despite the state not changing from `.cash(.collectingCash)` at any point.
- At other times, if we start a new order, and attempt to pay with card, it seemed to need a few attempts to work. One time I also saw the error `CardPresentPaymentInvalidatablePaymentOrchestratorError`, then it worked normally.
- I'm not quite sure when I should use `cancelCardReaderPreparation()` versus just `cardPresentPaymentService.cancelPayment()`. I went ahead with the later so we do not cancel `startPaymentOnCardReaderConnection` but it might have more sense to cancel the whole thing when a cash payment is initiated, and then restart it again when is successful or cancelled.

There's a specific issue logged for handling this on https://github.com/woocommerce/woocommerce-ios/issues/14682 , we can either continue in this PR, or I can revert the last change and move it there, whatever you prefer.

iamgabrielma replied:
> we can either continue in this PR, or I can revert the last change and move it there, whatever you prefer.

Just a quick update: From the sync call today it seems that we're going to still allow card events while the cash view is open without needing to navigate back to the totals view. I'll open a thread in slack to follow-up, then we can revert/clean-up this bit.

joshheald replied:
In Slack, Jirka and I discussed the option of keeping card payments open while Cash is being entered as above.

To do that we agreed that the minimum required would be to cancel the card payment as soon as `Mark payment complete` is tapped in the cash flow. I think you saw it but I can't see that code here; you mentioned you found it difficult though, did you try it and have to remove it?

joshheald commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -34,7 +31,7 @@ struct PointOfSaleCollectCashView: View {
             HStack {
                 Button(action: {
-                    dismiss()
+                    posModel.addMoreToCart()
```
We shouldn't take the merchant all the way back to the item list if they go back from cash payment – they may just want to take a card payment instead.

I think this should be something else, `posModel.cancelCashPayment()`, which would take the merchant back to the `finalizing` step, and restart the card reader preparation if it's connected.

iamgabrielma replied:
True, updated on https://github.com/woocommerce/woocommerce-ios/pull/14795/commits/d01819a62e4b88541e2e83b3dc8cf2728c462019. At this point we do not need to set `.finalizing` specifically, as it's already set from checking out, so switching payment state to the initial state is enough to navigate the user back. Added some tests as well 👍

joshheald commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -73,8 +70,7 @@ struct PointOfSaleCollectCashView: View {
                         try await markComplete()
-                        onCashPaymentSuccess()
-                        dismiss()
+                        posModel.cashPaymentSuccess()
```
`markComplete` just calls `posModel.collectCashPayment` anyway – as mentioned above, just remove this line and have the `collectCashPayment` function update the payment state when it completes.

joshheald reviewed (commented):
Nice one getting the state updated to reflect reality! Generally looking better – some comments in line about improvements to the state management.

The main thing left to do is cancelling and re-instating card reader preparation. Comments inline for details.

You may also want to try some `transition` modifiers for moving between cash payment and cash success... or grouped geometry to have the title and buttons appear to move around... it'll need some experimentation, but it doesn't feel right to just jump from one to another.

Comments inline to answer most of the other questions...

iamgabrielma committed:
make func private and move it to the model

iamgabrielma committed:
cancelling cash payment brings user back to payment selection

iamgabrielma committed:
Disallow card events while collecting cash

iamgabrielma committed:
restore state idle upon cancel. make test compile

iamgabrielma committed:
revert discarding card events while collecting cash

iamgabrielma committed:
isCashPaymentInProgress flag to intercept card events

Each time we tap the card on the reader, the .card payment state would override the current .cash payent state, changing the views. With this flag we filter card events when a cash transaction is in progress, so we can ignore them.

Allowing card payments while cash hasn’t been “marked as complete” yet, and disallowing doubles charges on the success screen.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/234">21.4</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -295,10 +308,17 @@ private extension PointOfSaleAggregateModel {
+
+                if self.isCashPaymentInProgress, case .card = newPaymentState {
+                    cardPresentPaymentService.cancelPayment()
+                    return .cash(.paymentSuccess)
```
Overriding the state here with `.cash(.paymentSuccess)` doesn't feel quite right and too optimistic, but since this only can happen when we're already showing the "success" view, I don't see much problem returning it 🤔 Combining this filter with the `isCashPaymentInProgress` flag is the only way I could find to intercept card events and cancel the payment attempt.

joshheald replied:
Soo... this exposed an issue with duplicating state (`isCashPaymentInProgress`). 

When testing, the first time that I tapped a card at the same time as tapping `mark payment complete`, I got a blank screen after seeing the success flash up.

I exited POS (my only option) and came back in, but `isCashPaymentInProgress` was still true then, and so every attempt to take a card payment after that failed because it was immediately cancelled (I think.)

Not something to go and separately fix – this is a symptom of issues with the approach of keeping separate flags for what's going on in the aggregate model; they can get out of sync and cause strange bugs.

joshheald replied:
Yeah, it's more than too optimistic, it's flawed logic, as it shows the success screen even if payment hasn't happened:

1. go to checkout using an error amount
2. connect the reader
3. tap collect cash payment but don't enter anything
4. tap a card on the reader

You'll see the cash success screen (which is incorrect anyway), and then it'll go in to card payment, which will fail because it gets cancelled. Then it goes to a blank screen which you can't navigate away from without exiting POS. The order doesn't get marked as paid, but the merchant has seen a success screen.

This can also be reproduced using a non-error amount, if the first card tap doesn't register correctly – if you kind of bring it in towards the reader slowly from far away, so it asks you to tap again.

iamgabrielma commented:
This is ready for re-review @joshheald 🙇 

As mentioned in slack, the only moment where we cancel payment attempts in within the success view, at other times it is possible to process a card payment if card is tapped. I'm wondering if calling `cancelCardReaderPreparation` is better, but this ends switching the payment state to `.card(.idle)` which messes up the payment flow, or perhaps I missed a better way to handle it.

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -129,6 +130,7 @@ extension PointOfSaleAggregateModel {
     }
 
     func startNewCart() {
+        isCashPaymentInProgress = false
```
Let's avoid adding flags like this; they tend to cause bugs as they're a duplication of state. We should rely on the `paymentState` to know whether a cash payment is in progress... if that doesn't work with the current payment state setup, the setup needs to be improved.

joshheald replied:
Generally speaking, you can either just switch over the state (which prompts you to handle every possibility, so helps you code for them), or add a computed var to the state enum to avoid these flags.

joshheald reviewed (commented):
Comments inline, but the card payments during cash entry doesn't work at the moment, and if the shopper does it, there's a good chance that the POS is left in a broken, unrecoverable state.

Despite yesterday's call, given the difficulty here I think we should opt for the most robust, straightforward approach – cancel the card payment before starting cash payment, and restart the card payment if cash is cancelled.

I could be wrong, so I'm going to check the approach quickly now, but by doing that we should be able to rule out a lot of potential states that we otherwise have to code for, test, and maintain.

joshheald committed:
14602 Cancel payment when order marked complete

Adds an async `cancelPayment` function to the CardPresentPaymentFacade. This isn’t as comprehensive as the fire-and-forget version, which remains, as it can’t cancel if it’s called during non-payment events, e.g. connection and searching.

We use the new async cancel to stop card payments from happening when a cash payment is accepted; it happens right before we mark the order complete.

Currently failure to cancel is unhandled; it would be good to improve that!

The state representation is incomplete with this approach. We should keep hold of the card payment state, if it’s still live in the background… and maybe even if it’s not. I’ve made suggestions for this in the comments.

Another (maybe better?) option (commented out in this commit) is to cancel card payments as soon as the cash option is selected.

iamgabrielma committed:
Disable card payment events as soon as cash collection starts

The optional try is necessary at this point, as if we do-catch then subsequent taps on “cash payment” where the cancelPayment always fails (because doesn’t exist) would land us in the catch block and don’t allow for collection. We can improve error handling here later on.

iamgabrielma committed:
Log error if cancel payment fails

failure to cancel the payment shouldn’t stop from continue with reader disconnection, so there’s no need to throw but we can log the error for future debugging purposes

iamgabrielma commented on WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentService.swift:
```diff
@@ -104,7 +104,11 @@ final class CardPresentPaymentService: CardPresentPaymentFacade {
+        do {
+            try await cancelPayment()
+        } catch {
+            DDLogError("Attempting to cancel the payment has failed \(error)")
```
We can log the `.couldNotCancelPayment` case here, but I don't think we want to throw, right? Both success or failure when cancelling the payment should still permit us to attempt reader disconnection.

joshheald replied:
Yeah, probably. Were you able to test it?

iamgabrielma commented on WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentService.swift:
```diff
@@ -161,6 +165,18 @@ final class CardPresentPaymentService: CardPresentPaymentFacade {
     }
+
+    @MainActor
+    func cancelPayment() async throws {
+        try await withCheckedThrowingContinuation { continuation in
+            var nillableContinuation: CheckedContinuation<Void, any Error>? = continuation
+            let action = CardPresentPaymentAction.cancelPayment { result in
+                nillableContinuation?.resume(with: result)
+                nillableContinuation = nil
+            }
+            stores.dispatch(action)
+        }
+    }
```
I understand the use of `nillableContinuation` here is just so we assure that we're not calling the resume more than once?

joshheald replied:
Exactly. 

It's not a fix for everything; ideally we'd be able to prevent further `cancelPayment` actions from being dispatched, not just prevent ourselves from resuming more than once... but that's not really possible, it'll always be non-nil each time dispatch is called. We just swallow issues where the completion handler from a single call is called more than once.

It might not be needed, I do it as a bit of a reflex now.

iamgabrielma commented:
Thanks for your help with this @joshheald , and for your time during the call 🙇 I've taken your commit and went ahead with the solution for "cancelling as soon as we enter in cash collection", as seems much more adequate than just waiting until is "marked as complete". 

Along this I've also added some simple error logging on https://github.com/woocommerce/woocommerce-ios/pull/14795/commits/2987449e0d2a642f1109bc79ce252b24dae1a9da and removed the comment with changes for the payment state representation, as now we're cancelling it all the time and we do not keep it live in the background. As well as made the start/cancel methods [async](https://github.com/woocommerce/woocommerce-ios/pull/14795/commits/746f37cf68e86061cb0422296bd446fbbe539da3) so it pauses execution until cancellation returns. 

As mentioned in Slack, at some point I was able to crash the system after a payment failure. I haven't been able to replicate it since then, so I could have been just because I had the debugger enabled. Just in case, the flow I followed was Connect reader > Cash Payment > cancel cash payment > attempt to collect via card > failed > try again > failed and crash (possibly I tried 3 or more times before the crash). With this error:

```
Error taking payment: alreadyRetried(Hardware.CardReaderServiceError.paymentCaptureWithPaymentMethod(underlyingError: Hardware.UnderlyingError.paymentDeclinedByPaymentProcessorAPI(declineReason: Hardware.DeclineReason.fraud), paymentMethod: Hardware.PaymentMethod.cardPresent(details: Hardware.CardPresentTransactionDetails(last4: "9969", expMonth: 3, expYear: 2023, cardholderName: Optional("CARDHOLDER/VISA"), brand: Hardware.CardBrand.visa, generatedCard: nil, receipt: Optional(Hardware.ReceiptDetails(applicationPreferredName: "Stripe Credit", dedicatedFileName: "A000000003101001", authorizationResponseCode: "3031", applicationCryptogram: "E9D16FE3DBA02C49", terminalVerificationResults: "0000000000", transactionStatusInformation: "0000", accountType: Optional("credit"))), emvAuthData: Optional("8A023031"), wallet: Optional(Hardware.Wallet(type: nil)), cardBrandAssociatedIdentifier: Optional(0.0)))))
```
The rejection reason as "fraud" throw me off a bit here, maybe due to the testing amount? After this I've tried several times and it works just fine 🤔

iamgabrielma committed:
update start and cancel cash payment signatures to be async

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -201,14 +199,22 @@ extension PointOfSaleAggregateModel {
+    // Once we get the callback from the card service, we switch to cash collection state
+    @MainActor
+    func startCashPayment() async {
+        try? await cardPresentPaymentService.cancelPayment()
```
Is it worth handling the errors here? I think it probably is, but not totally sure how. Starting the cash payment doesn't seem like the way though...

Feel free to do this under a new issue.

iamgabrielma replied:
Logged: https://github.com/woocommerce/woocommerce-ios/issues/14858

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -201,14 +199,22 @@ extension PointOfSaleAggregateModel {
+    func cancelCashPayment() async {
+        if case .connected = cardReaderConnectionStatus {
+            try? await cardPresentPaymentService.cancelPayment()
+            await collectPayment()
```
It might be worth renaming this to `collectCardPayment` now, for clarity

iamgabrielma replied:
Agree! Done https://github.com/woocommerce/woocommerce-ios/pull/14795/commits/91b3b8d00b6bb2428d462e23759c999a5cb111a9

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -201,14 +199,22 @@ extension PointOfSaleAggregateModel {
+    @MainActor
+    func cancelCashPayment() async {
+        if case .connected = cardReaderConnectionStatus {
+            try? await cardPresentPaymentService.cancelPayment()
```
Why do we need to cancel the card payment again here?

iamgabrielma replied:
We don't! It would be redundant. I think might be there from the previous attempt where we were not yet cancelling on starting cash collection. Removed: https://github.com/woocommerce/woocommerce-ios/pull/14795/commits/6e4d54ff84ac04dc7684eab447a3635d2561d164

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -201,14 +199,22 @@ extension PointOfSaleAggregateModel {
+            try? await cardPresentPaymentService.cancelPayment()
+            await collectPayment()
+        } else {
+            paymentState = .card(.idle)
```
Perhaps `startPaymentWhenCardReaderConnected` might be a better alternative? Not sure, it works fine as is but you could experiment in a new issue.

iamgabrielma replied:
Logged https://github.com/woocommerce/woocommerce-ios/issues/14859

joshheald reviewed (approved):
Various small suggestions inline but it works well, so feel free to merge and address issues in future PRs.

I noticed the backgrould is wrong in dark mode:
![IMG_0276](https://github.com/user-attachments/assets/050241e6-7b67-4f45-9e03-6ab151d09d04)

joshheald commented:
> The rejection reason as "fraud" throw me off a bit here, maybe due to the testing amount? After this I've tried several times and it works just fine 🤔

It's just the message for one of the testing amounts 😊

iamgabrielma committed:
remove redundant call to cancelPayment

Since we now cancel card payments on starting the cash collection there’s no necessity to cancel card again if we cancel cash, it’s redudant.

iamgabrielma committed:
rename for clarity. make private

iamgabrielma commented:
Thanks again for the review and the pointers 🙇 

> I noticed the backgrould is wrong in dark mode:

Yep! This is already addressed on a [separate PR](https://github.com/woocommerce/woocommerce-ios/pull/14830#issuecomment-2586117573)
